### ISSUE TITLE
SQLiteDatabaseLockedException in 1.3.0-8

### ISSUE THREAD

**Issue by npomfret** [couchbase/couchbase-lite-android#929]
```
07-02 14:37:42.356 14726-18542/com.pomochat I/Router: Router: Sending longpoll response: START
07-02 14:37:42.367 14726-18542/com.pomochat I/Router: Router: Sending longpoll response: END
07-02 14:37:44.334 14726-14863/com.pomochat I/Router: Router: Sending longpoll response: START
07-02 14:37:44.344 14726-14863/com.pomochat I/Router: Router: Sending longpoll response: END
07-02 14:37:44.421 14726-19684/com.pomochat E/Batcher: com.couchbase.lite.support.Batcher@1309ab08: Error while waiting for pending futures
       java.util.concurrent.ExecutionException: com.couchbase.lite.internal.database.sqlite.exception.SQLiteDatabaseLockedException: database is locked (code 5)
       at java.util.concurrent.FutureTask.report(FutureTask.java:93)
       at java.util.concurrent.FutureTask.get(FutureTask.java:163)
       at com.couchbase.lite.support.Batcher.waitForPendingFutures(Batcher.java:256)
       at com.couchbase.lite.replicator.ReplicationInternal.waitBatcherCompleted(ReplicationInternal.java:1903)
       at com.couchbase.lite.replicator.PullerInternal.waitDownloadsToInsertBatcherCompleted(PullerInternal.java:1020)
       at com.couchbase.lite.replicator.PullerInternal.waitForAllTasksCompleted(PullerInternal.java:1015)
       at com.couchbase.lite.replicator.ReplicationInternal.waitForPendingFutures(ReplicationInternal.java:1858)
       at com.couchbase.lite.replicator.PullerInternal$14.run(PullerInternal.java:929)
       at java.lang.Thread.run(Thread.java:818)
    Caused by: com.couchbase.lite.internal.database.sqlite.exception.SQLiteDatabaseLockedException: database is locked (code 5)
       at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.nativeExecute(Native Method)
       at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.execute(SQLiteConnection.java:486)
       at com.couchbase.lite.internal.database.sqlite.SQLiteSession.beginTransactionUnchecked(SQLiteSession.java:330)
       at com.couchbase.lite.internal.database.sqlite.SQLiteSession.beginTransaction(SQLiteSession.java:305)
       at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.beginTransaction(SQLiteDatabase.java:498)
       at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.beginTransaction(SQLiteDatabase.java:409)
       at com.couchbase.lite.storage.SQLiteStorageEngineBase.beginTransaction(SQLiteStorageEngineBase.java:141)
       at com.couchbase.lite.store.SQLiteStore.beginTransaction(SQLiteStore.java:2296)
       at com.couchbase.lite.store.SQLiteStore.runInTransaction(SQLiteStore.java:711)
       at com.couchbase.lite.replicator.PullerInternal.insertDownloads(PullerInternal.java:562)
       at com.couchbase.lite.replicator.PullerInternal$2.process(PullerInternal.java:127)
       at com.couchbase.lite.support.Batcher.processNow(Batcher.java:396)
       at com.couchbase.lite.support.Batcher.access$300(Batcher.java:29)
       at com.couchbase.lite.support.Batcher$2.run(Batcher.java:330)
       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
       at java.lang.Thread.run(Thread.java:818) 
07-02 14:37:45.981 14726-15636/com.pomochat W/RemoteRequest: com.couchbase.lite.replicator.RemoteRequest {PUT, http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb}: Got error status: 409 for http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb. Reason: Conflict
07-02 14:37:45.982 14726-15636/com.pomochat I/Sync: com.couchbase.lite.replicator.ReplicationInternal$8@2c551102: could not save remote checkpoint: 409 CONFLICT
07-02 14:37:45.982 14726-15636/com.pomochat I/Sync: PullerInternal{unknown, pull, b1108}: Refreshing remote checkpoint to get its _rev...
07-02 14:37:45.988 14726-19683/com.pomochat W/RemoteRequest: com.couchbase.lite.replicator.RemoteRequest {PUT, http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb}: Got error status: 409 for http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb. Reason: Conflict
07-02 14:37:45.989 14726-19683/com.pomochat I/Sync: com.couchbase.lite.replicator.ReplicationInternal$8@5a0a649: could not save remote checkpoint: 409 CONFLICT
07-02 14:37:45.989 14726-19683/com.pomochat I/Sync: PullerInternal{unknown, pull, b1108}: Refreshing remote checkpoint to get its _rev...
07-02 14:37:45.990 14726-14865/com.pomochat W/RemoteRequest: com.couchbase.lite.replicator.RemoteRequest {PUT, http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb}: Got error status: 409 for http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb. Reason: Conflict
07-02 14:37:45.991 14726-14865/com.pomochat I/Sync: com.couchbase.lite.replicator.ReplicationInternal$8@d5c557c: could not save remote checkpoint: 409 CONFLICT
07-02 14:37:45.991 14726-14865/com.pomochat I/Sync: PullerInternal{unknown, pull, b1108}: Refreshing remote checkpoint to get its _rev...
07-02 14:37:46.059 14726-19291/com.pomochat W/RemoteRequest: com.couchbase.lite.replicator.RemoteRequest {PUT, http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb}: Got error status: 409 for http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb. Reason: Conflict
07-02 14:37:46.063 14726-19291/com.pomochat I/Sync: com.couchbase.lite.replicator.ReplicationInternal$8@3b1edb2d: could not save remote checkpoint: 409 CONFLICT
07-02 14:37:46.063 14726-19291/com.pomochat I/Sync: PullerInternal{unknown, pull, b1108}: Refreshing remote checkpoint to get its _rev...
0
```

From there it looks like replication is stuck, it doesn't get beyond this state:

```
 { completed_change_count: 0,
                                                               source: 'http://192.168.1.185:4986/sync_gateway',
                                                               task: 'repl010',
                                                               status: 'Processed 0 / 9 changes',
                                                               progress: 0,
                                                               continuous: true,
                                                               target: 'pc_50_063df152-8fff-4aa3-beb1-e63849155460',
                                                               change_count: 9,
                                                               type: 'Replication' }
```

I also see a lot of errors like this:

`07-02 14:48:02.216 14726-14865/com.pomochat W/RemoteRequest: com.couchbase.lite.replicator.RemoteRequest {PUT, http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb}: Got error status: 409 for http://192.168.1.185:4986/sync_gateway/_local/b1108e0df5ad9bc46f8074ccf05e28f5aaca66bb. Reason: Conflict
`

**Comment by hideki** [comment]
Hi @npomfret,
We would like to reproduce this issue here. Could you please provide detail steps to reproduce the problem?
Thanks,

**Comment by hideki** [comment]
@sethrosetter,
Have you ever seen this issue with functional test or robot test?

**Comment by npomfret** [comment]
No sorry.  It just happened the once.  Or at least I've only noticed it once.

**Comment by hideki** [comment]
I believe this issue is related with #902.

**Comment by npomfret** [comment]
It happens fairly regularly with 1.3.0-37

Here's some logs around the time it happened last https://gist.github.com/npomfret/a2b419360193eb8e5f113578ea74098b

**Comment by hideki** [comment]
Hi @npomfret,
Could you please provide Android version or API version that you are using for your test?

**Comment by npomfret** [comment]
Android 5.1

**Comment by pasin** [comment]
SQLiteDatabaseLockedException (code 5) normally happens when a SQLite connection is shared between 2 threads. As Android SQLiteDatabase Connection is local thread bound, I'm wondering how this could happen. Maybe a race condition when acquiring a SQLite connection?

**Comment by hideki** [comment]
From comments in source codes, `SQLiteDatabaseLockedException(SQLITE_BUSY)` could happen with multiple connection with Write-Ahead Log (WAL).

For CBL Android, we might need to increase `BUSY_TIMEOUT_MS` value.

```java
/* Busy timeout in milliseconds.
 * If another connection (possibly in another process) has the database locked for
 * longer than this amount of time then SQLite will generate a SQLITE_BUSY error.
 * The SQLITE_BUSY error is then raised as a SQLiteDatabaseLockedException.
 *
 * In ordinary usage, busy timeouts are quite rare.  Most databases only ever
 * have a single open connection at a time unless they are using WAL.  When using
 * WAL, a timeout could occur if one connection is busy performing an auto-checkpoint
 * operation.  The busy timeout needs to be long enough to tolerate slow I/O write
 * operations but not so long as to cause the application to hang indefinitely if
 * there is a problem acquiring a database lock.
 */
static const int BUSY_TIMEOUT_MS = 2500;
```
https://github.com/couchbase/couchbase-lite-java-native/blob/master/jni/source/com_couchbase_lite_internal_database_sqlite_SQLiteConnection.cpp#L24-L36

```java
    // Set the default busy handler to retry automatically before returning SQLITE_BUSY.
    err = sqlite3_busy_timeout(db, BUSY_TIMEOUT_MS);
    if (err != SQLITE_OK) {
        throw_sqlite3_exception(env, db, "Could not set busy timeout");
        sqlite3_close(db);
        return 0;
    }
```
https://github.com/couchbase/couchbase-lite-java-native/blob/master/jni/source/com_couchbase_lite_internal_database_sqlite_SQLiteConnection.cpp#L107-L113

**Comment by hideki** [comment]
Or, as `SQLiteDatabaseLockedException` is always thrown from `beginTransaction()`, so we might need to retry `beginTransaction()` if exception is `SQLiteDatabaseLockedException`

>Thrown if the database engine was unable to acquire the database locks it needs to do its job. If the statement is a [COMMIT] or occurs outside of an explicit transaction, then you can retry the statement. If the statement is not a [COMMIT] and occurs within a explicit transaction then you should rollback the transaction before continuing.

https://developer.android.com/reference/android/database/sqlite/SQLiteDatabaseLockedException.html

**Comment by pasin** [comment]
2.5 seems do be long enough in normal cases but giving that I don't really know about auto-checkpoint operation so it may not be enough for slow devices or android emulators.

@hideki , please see this.
https://github.com/couchbase/couchbase-lite-ios/blob/6434af8f71d8962872d0a835b5acc2fd9f611f46/Source/CBL_SQLiteStorage.m#L601-L633

**Comment by hideki** [comment]
Thanks for link to iOS implementation. From iOS implementation and also JavaDoc of `SQLiteDatabaseLockedException`,  Retry is correct approach.

**Comment by hideki** [comment]
I tried to update `runInTransaction()` method of SQLiteStore.java, and apply it to all places that uses `beginTransaction()` and `endTransaction()`. However, there are many codes should be updated. I feel risk to fix the issue with this approach for 1.3.0 release. Instead, I am considering to fix this by applying retry logic in `beginTransaction()` for v1.3.0 release as currently we can observe this Exception only in `beginTransaction()`. From SQLite documentation, automatic checkpoint occurs when COMMIT occurs. As far as I know, SQLiteStore write operations are executed in transaction. Updating `runInTransaction` method and applying it to all codes that uses transaction will be done for next release (1.4.0 ??).

> Automatic Checkpoint
By default, SQLite will automatically checkpoint whenever a COMMIT occurs that causes the WAL file to be 1000 pages or more in size, or when the last database connection on a database file closes. The default configuration is intended to work well for most applications. But programs that want more control can force a checkpoint using the wal_checkpoint pragma or by calling the sqlite3_wal_checkpoint() C interface. The automatic checkpoint threshold can be changed or automatic checkpointing can be completely disabled using the wal_autocheckpoint pragma or by calling the sqlite3_wal_autocheckpoint() C interface. A program can also use sqlite3_wal_hook() to register a callback to be invoked whenever any transaction commits to the WAL. This callback can then invoke sqlite3_wal_checkpoint() or sqlite3_wal_checkpoint_v2() based on whatever criteria it thinks is appropriate. (The automatic checkpoint mechanism is implemented as a simple wrapper around sqlite3_wal_hook().)

https://www.sqlite.org/wal.html

**Comment by pasin** [comment]
CC: @sethrosetter

**Comment by hideki** [comment]
Hi @npomfret,

When you have time, could you please try the build 1.3.0-43 which has fix for your error scenario?
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-java-core/1.3.0-43/

Thanks!

**Comment by npomfret** [comment]
Looks much better so far - will keep monitoring over the next few days.  Thanks!

**Comment by hideki** [comment]
@npomfret Thank you for testing this.

**Comment by npomfret** [comment]
I got this today with v 1.3.0-45 - sorry I didn't manage to grab the full log but will look out for it again.

```
07-31 15:23:31.760 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#1)...
07-31 15:23:33.929 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#2)...
07-31 15:23:36.099 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#3)...
07-31 15:23:38.269 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#4)...
07-31 15:23:40.439 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#5)...
07-31 15:23:42.609 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#6)...
07-31 15:23:44.783 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#7)...
07-31 15:23:46.949 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#8)...
07-31 15:23:49.119 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#9)...
07-31 15:23:51.289 21805-11531/com.pomochat I/Database: Db busy, retrying transaction (#10)...
07-31 15:23:53.458 21805-11531/com.pomochat E/Database: Db busy, too many retries, giving up
07-31 15:23:53.496 21805-11531/com.pomochat W/System.err: com.couchbase.lite.internal.database.sqlite.exception.SQLiteDatabaseLockedException: database is locked (code 5)
07-31 15:23:53.496 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.nativeExecute(Native Method)
07-31 15:23:53.497 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.execute(SQLiteConnection.java:486)
07-31 15:23:53.497 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.internal.database.sqlite.SQLiteSession.beginTransactionUnchecked(SQLiteSession.java:330)
07-31 15:23:53.498 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.internal.database.sqlite.SQLiteSession.beginTransaction(SQLiteSession.java:305)
07-31 15:23:53.498 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.beginTransaction(SQLiteDatabase.java:499)
07-31 15:23:53.499 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.beginTransaction(SQLiteDatabase.java:410)
07-31 15:23:53.499 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.storage.SQLiteStorageEngineBase.beginTransaction(SQLiteStorageEngineBase.java:150)
07-31 15:23:53.503 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.store.SQLiteStore.beginTransaction(SQLiteStore.java:2309)
07-31 15:23:53.503 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.store.SQLiteStore.open(SQLiteStore.java:235)
07-31 15:23:53.504 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.Database.open(Database.java:1251)
07-31 15:23:53.505 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.Database.open(Database.java:1185)
07-31 15:23:53.506 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.router.Router.openDB(Router.java:345)
07-31 15:23:53.506 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.router.Router.start(Router.java:474)
07-31 15:23:53.507 21805-11531/com.pomochat W/System.err:     at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
```

**Comment by hideki** [comment]
ref: https://github.com/couchbase/couchbase-lite-android/issues/950

**Comment by hideki** [comment]
SQLite library waits for 2.5sec if DB is busy state. By fix of this issue, now CBL v1.3.0 waits maximum 25 sec (2.5 sec x 10 times retries). It seems 25sec is not enough in some case.

We will keep tracking this DBBUSY issue with #950
